apps/api-extractor/src/extractor/Extractor.ts

```diff
@@ -164,11 +164,50 @@ export class Extractor {
    * @param options - IExtractor options.
    */
   public static processProjectFromConfigFile(jsonConfigFile: string, options?: IExtractorOptions): void {
-    const configObject: IExtractorConfig = JsonFile.loadAndValidate(jsonConfigFile, Extractor.jsonSchema);
+    const configObject: IExtractorConfig = this.loadConfigObject(jsonConfigFile);
     const extractor: Extractor = new Extractor(configObject, options);
     extractor.processProject();
   }
 
+  /**
+   * Loads the api extractor config file in Extractor Config object.
+   * @param jsonConfigFile - Path to api extractor json config file.
+   */
+  public static loadConfigObject(jsonConfigFile: string): IExtractorConfig {
+    // Set to keep track of config files which have been processed.
+    const pathSet: Set<string> = new Set<string>();
+    // Get absolute path of config file.
+    let currentConfigFilePath: string = path.resolve(jsonConfigFile);
+    pathSet.add(currentConfigFilePath);
+
+    let extractorConfig: IExtractorConfig = JsonFile.load(jsonConfigFile);
+
+    while (extractorConfig.extends) {
+      // Populate the api extractor config path defined in extends relative to current config path.
+      currentConfigFilePath = path.resolve(path.dirname(currentConfigFilePath), extractorConfig.extends);
+
+      // Check if this file was already processed.
+      if (pathSet.has(currentConfigFilePath)) {
+        throw new Error('The api extractor config files contains a cycle. '
+        + currentConfigFilePath + ' is being processed again. '
+        + 'Please check the extends values in config files.');
+      }
+      pathSet.add(currentConfigFilePath);
+
+      // Remove extends property from config for current config.
+      delete extractorConfig.extends;
+
+      // Load the extractor config defined in extends property.
+      const baseConfig: IExtractorConfig = JsonFile.load(currentConfigFilePath);
+      lodash.merge(baseConfig, extractorConfig);
```

Comment:
>merge [](start = 13, length = 5)

I'm not sure we want to recursively merge every object in the config. Some we may want to completely replace. @pgonzal?

Reply:
What are the lodash.merge() semantics?  In particular, how are arrays and nested objects handled?

---
In reply to: [236477963](https://github.com/Microsoft/web-build-tools/pull/957#discussion_r236477963) [](ancestors = 236477963)